packages/filesystem/src/browser/filesystem-save-resource-service.ts

```diff
@@ -0,0 +1,130 @@
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License v. 2.0 which is available at
+// http://www.eclipse.org/legal/epl-2.0.
+//
+// This Source Code may also be made available under the following Secondary
+// Licenses when the conditions for such availability set forth in the Eclipse
+// Public License v. 2.0 are satisfied: GNU General Public License, version 2
+// with the GNU Classpath Exception which is available at
+// https://www.gnu.org/software/classpath/license.html.
+//
+// SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+// *****************************************************************************
+
+import { environment, nls, UNTITLED_SCHEME } from '@theia/core';
+import { inject, injectable } from '@theia/core/shared/inversify';
+import { Navigatable, Saveable, SaveableSource, SaveOptions, Widget, open, OpenerService, ConfirmDialog, FormatType, CommonCommands } from '@theia/core/lib/browser';
+import { SaveResourceService } from '@theia/core/lib/browser/save-resource-service';
+import URI from '@theia/core/lib/common/uri';
+import { FileService } from './file-service';
+import { FileDialogService } from './file-dialog';
+
+@injectable()
+export class FilesystemSaveResourceService extends SaveResourceService {
+
+    @inject(FileService) protected readonly fileService: FileService;
+    @inject(FileDialogService) protected readonly fileDialogService: FileDialogService;
+    @inject(OpenerService) protected readonly openerService: OpenerService;
+
+    /**
+     * This method ensures a few things about `widget`:
+     * - `widget.getResourceUri()` actually returns a URI.
+     * - `widget.saveable.createSnapshot` is defined.
+     * - `widget.saveable.revert` is defined.
+     */
+    override canSaveAs(widget: Widget | undefined): widget is Widget & SaveableSource & Navigatable {
+        return widget !== undefined
+            && Saveable.isSource(widget)
+            && typeof widget.saveable.createSnapshot === 'function'
+            && typeof widget.saveable.revert === 'function'
+            && Navigatable.is(widget)
+            && widget.getResourceUri() !== undefined;
+    }
+
+    /**
+     * Save `sourceWidget` to a new file picked by the user.
+     */
+    override async saveAs(sourceWidget: Widget & SaveableSource & Navigatable, options?: SaveOptions): Promise<void> {
+        let exist: boolean = false;
+        let overwrite: boolean = false;
+        let selected: URI | undefined;
+        const canSave = this.canSaveNotSaveAs(sourceWidget);
+        const uri: URI = sourceWidget.getResourceUri()!;
+        let stat;
+        if (uri.scheme === 'file') {
+            stat = await this.fileService.resolve(uri).catch(() => undefined);
+        } else if (uri.scheme === UNTITLED_SCHEME) {
+            stat = await this.fileService.resolve(uri.withScheme('file').parent).catch(() => undefined);
```

Comment:
I noticed that this usually leads to an undefined `stat` variable. At least it does that in my test extension.

Suggestion: Use the `UntitledFileLocationProvider#getParent()` method, if the uri parent returns a root path (`uri.withScheme('file').parent.path === '/'` or something similar), to get a better starting point for the file dialog.

Reply:
I've eliminated the check for a `stat` here in favor of just letting the `FileDialogService` use the `UntitledfileLocationProvider#getParent` logic (now `UserWorkingDirectoryProvider).